src/basekernels/matern.jl

```diff
@@ -37,14 +37,15 @@ MaternKernel(; nu::Real=1.5, ν::Real=nu, metric=Euclidean()) = MaternKernel(ν,
 
 @functor MaternKernel
 
-@inline function kappa(k::MaternKernel, d::Real)
-    result = _matern(only(k.ν), d)
-    return ifelse(iszero(d), one(result), result)
-end
+@inline kappa(k::MaternKernel, d::Real) = _matern(only(k.ν), d)
 
 function _matern(ν::Real, d::Real)
-    y = sqrt(2ν) * d
-    return exp((one(d) - ν) * logtwo - loggamma(ν) + ν * log(y) + log(besselk(ν, y)))
+    if iszero(d)
+        return one(d)
```

Comment:
I think this is problematic as it very easily creates type stability issues.

Reply:
Maybe something like
```suggestion
        return one(float(Base.promote_typeof(ν, d)))
```
would be sufficient.

Reply:
Maybe a better approach though (also e.g. regarding the computation in the other branch) would be to define
```julia
_matern(ν::Real, d::Real) = _matern(promote(ν, d)...)
function _matern(ν::T, d::T) where {T<:Real}
....
```

Reply:
Good point

Reply:
Promotion + ForwardDiff / ReverseDiff don't appear to play nicely. Neither of the above patterns seems to work unfortunately.

Reply:
I don't believe the new implementation to be any _worse_ than the previous, so unless you strongly object, I would like to overlook this for now.

Reply:
Ah that's unfortunate and a bit surprising. I'm fine with keeping the PR as it is and to improve it at some point.